### ISSUE TITLE
[Due for payment 2025-02-18] [$250] Add distance label on map route

### ISSUE THREAD

**Issue by dubielzyk-expensify** [Expensify/App#53421]
## Problem

Today when users submit a distance expense there's no way for users to see how long the distance is without continuing.

## Solution

To improve quality of life and bring clarity to the distance request, let's add a distance label on the map route to ensure users can see the distance before they go to the detail screen.

![Image](https://github.com/user-attachments/assets/5db640ff-dc0b-4219-bd79-e462a42329a5)

I believe the Mapbox API allows us to do [this pretty easily](https://stackoverflow.com/questions/53772389/how-can-i-display-distance-markings-on-a-line-mapbox-gl) but I'm open to styling it differently depending on what controls we have. 

- If the route contains multiple stops, we should not do individual legs, but rather only the distance of the full route.
- If your default workspace is the personal workspace, we'd display the unit associated with the P2P rate/unit we've assigned you from the [currencyToDefaultMileageRate](https://github.com/Expensify/Web-Expensify/blob/4710c6da5e089ce7ad59575901c2b6300c3ab1c6/_constant.php#L317-L974) list based on the output currency of the personal workspace.
- If you tap/click on the tooltip it changes from miles/km, so everyone has the chance to see both if they want (i.e travelling abroad or something).

cc @Expensify/design 
<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~021864707242087404252</li>
        <li>Upwork Job ID: 1864707242087404252</li>
        <li>Last Price Increase: 2024-12-26</li>
<li>Automatic offers: </li>
<ul>
<li>shubham1206agra | Contributor | 105535036</li>
<li>truph01 | Contributor | 105692875</li>
</ul></li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @shubham1206agra</details>

**Comment by truph01** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
- Add distance label on map route

### What is the root cause of that problem?
- New feature
### What changes do you think we should make in order to solve the problem?
- We need to create a custom symbol component based on design. This component will get the distance data from transaction (via ```TransactionUtils.getDistanceInMeters(transaction, 'km')```), and then display that distance data in [here](https://github.com/Expensify/App/blob/0eec30d00c131d3f9caacf1de6afd33c83bc351e/src/components/MapView/MapViewImpl.website.tsx#L259), like we did when display the current position marker. 

1. Get the distance data.

- As mentioned above, the distance data can be get from ```TransactionUtils.getDistanceInMeters(transaction)```).

2. Get the coordinate to display the distance label.

- The coordinate of the distance label can be the middle element in the directionCoordinates. So in [here](https://github.com/Expensify/App/blob/0eec30d00c131d3f9caacf1de6afd33c83bc351e/src/components/MapView/MapViewImpl.website.tsx#L162), we can add:

```
        const distanceSymbolCoorinate = useMemo(() => {
            const length = directionCoordinates?.length;
            // If the array is empty, return undefined
            if (!length) return undefined;

            // Find the index of the middle element
            const middleIndex = Math.floor(length / 2);

            // Return the middle element
            return directionCoordinates.at(middleIndex);
        }, [directionCoordinates]);
``` 


3. Display the distance label.

- In [here](https://github.com/Expensify/App/blob/0eec30d00c131d3f9caacf1de6afd33c83bc351e/src/components/MapView/MapViewImpl.website.tsx#L259), use the Marker to display the distance label:

```
                    {distanceSymbolCoorinate && (
                        <Marker
                            key={'distance'}
                            longitude={distanceSymbolCoorinate.at(0) ?? 0}
                            latitude={distanceSymbolCoorinate.at(1) ?? 0}
                        >
                            <View style={{marginRight: 100}}>
                                <View
                                    style={{backgroundColor: '#60d184', paddingHorizontal: 10, paddingVertical: 4, color: 'white', borderRadius: 20, fontWeight: 'bold', textAlign: 'center'}}
                                >{`${distance} km`}</View>
                            </View>
                        </Marker>
                    )}
```
It is just the mock style. The detailed should be address when we implement in PR.


4. The similar can be applied in native.
### What specific scenarios should we cover in automated tests to prevent reintroducing this issue in the future?

### What alternative solutions did you explore? (Optional)



**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by dubielzyk-expensify** [comment]
Sorry @JmillsExpensify just saw your concern about whether to use miles or km. Do we wanna do something like this instead? I assume adding metric/imperial preference in profile is too much at this point?

![CleanShot 2024-12-10 at 13 39 35@2x](https://github.com/user-attachments/assets/c180eacb-902d-4f4e-8fc2-bdea1e8c3601)

cc @Expensify/design

**Comment by shawnborton** [comment]
I quite like your middle treatment there.

Is there a way to just use someone's primary currency/location to determine whether we initially show km or mi? Most of our customers are US-based and don't need to worry about km so I don't want to over solve this problem, so-to-speak.

**Comment by dannymcclain** [comment]
>I quite like your middle treatment there.
>Is there a way to just use someone's primary currency/location to determine whether we initially show km or mi? Most of our customers are US-based and don't need to worry about km so I don't want to over solve this problem, so-to-speak.

Big agree with all of this. Love the middle treatment. I think we should either make an educated guess about which to make the primary unit, or just always lead with miles and let kilometers be the secondary for now (since most of our customers are US-based).

**Comment by dubielzyk-expensify** [comment]
I'll add the question to the original post and see if there's a way we can use timezone to detect 👍 Let's use the middle one

**Comment by dannymcclain** [comment]
Can you verify the background color you're using for the background of the little distance label? It should be the same color as the route, but in your video/screenshot it looks a little different.

Also I think the smaller distance (on the bottom) should be using a different text color (`green100` from the looks of the screenshot in OP).

**Comment by Gonals** [comment]
@truph01 all yours!

**Comment by dubielzyk-expensify** [comment]
Background color should be the same as the route color. We generally use our `success` color variable which is: `#03D47C`

Here's the specs for the callout:

![Image](https://github.com/user-attachments/assets/b4e5eb9c-d57b-4cd3-819f-00b23f3aaa3b)

Type settings:
Size `label` which is `Expensify Neue, Bold, 13px `
Color should be `#fff`

**Comment by shawnborton** [comment]
For the color, are we able to just pull the same color variable we use for button text? I say that because I don't think we use a pure white color for button labels.

**Comment by dubielzyk-expensify** [comment]
Yeah, we can use the text color, but it looks quite muddy. I dunno it feels worse than it does on our buttons. Happy to use text color:

![Image](https://github.com/user-attachments/assets/25dbcc7b-5b96-40a9-8e9c-a687e81551c3)

Which would be our text color variable or `#E7ECE9`

**Comment by shawnborton** [comment]
Sorry, I don't mean text color, I mean buttonText color.

**Comment by dubielzyk-expensify** [comment]
Editing my comment as we speak haha. Let's use the text color variable (`#FCFBF9`).

**Comment by shawnborton** [comment]
Our button text isn't a pure #FFF white:

![Image](https://github.com/user-attachments/assets/927fae95-71e2-4be1-b8b4-24db859491e9)

But it is lighter than dark mode text for instance.

**Comment by shawnborton** [comment]
Haha right on, same page!

**Comment by dubielzyk-expensify** [comment]
> Based on your design, all the text will have the same font size and font weight, right?

Yeah. Given we don't know if they use miles or kilometers, I think it's best to not be partial to either. 

Can you show me a video of how this looks from entering the start -> stop and interacting with the map?

**Comment by shawnborton** [comment]
I wonder if we can choose mi or km based on the location of the map? For instance, a good majority of our users are USD-based, and have no use for the km value... Happy to do nothing for now, just noting that we'd be showing unnecessary information to a majority of our users.

**Comment by dubielzyk-expensify** [comment]
Ohh I love that. Maybe Mapbox has something clever here? Keen to hear what can be done cause I agree that we should ideally be showing the local unit

**Comment by trjExpensify** [comment]
Here's an idea... 

- Show the distance on the map in the unit of your primary workspace (activePolicyID) if it's a group workspace.
     - If your default workspace is the personal workspace, we'd display the unit associated with the P2P rate/unit we've assigned you from the [currencyToDefaultMileageRate](https://github.com/Expensify/Web-Expensify/blob/4710c6da5e089ce7ad59575901c2b6300c3ab1c6/_constant.php#L317-L974) list based on the output currency of that workspace.
- If you tap/click on the tooltip it changes from miles/km, so everyone has the chance to see both if they want (i.e travelling abroad or something). 

How I think that would work overall is as follows:

- If you use `Create expense` in global create (where a recipient isn't selected yet), we display the unit of your default workspace (activePolicyID) || extrapolated from the personal workspace's output currency. 
   - Extra: If you then select a workspace that's different to your default workspace, we can change the unit we display on the confirmation screen preview to match that workspace's setting perhaps? 
- If you create the expense from within a specific workspace chat, thus the unit to track distance on the workspace is known, we display that unit.
- If you create the expense from the quick action button, same as above. We know the recipient and which unit we'll be tracking in.

**Comment by dannymcclain** [comment]
I think we should avoid trying to add any kind of click interaction to the tooltip—that seems like it will just lead to problems (and would be pretty hidden/unintuitive anyways). But I am in favor of trying to determine which option should be more prominent, even if it's just mostly a guess.

**Comment by trjExpensify** [comment]
Kewl, not passionate about that. Just thought it was quite nice if you could change it.

**Comment by shawnborton** [comment]
I like the idea Tom, and I don't feel strongly about the tappable tooltip or not. 

That being said, my active policy is an Expensify US policy, but everytime I go to create an expense while in Spain, it defaults to EUR. So maybe the functionality you are referencing is actually not working correctly at the moment? Either way, I like where you are going with it.

**Comment by trjExpensify** [comment]
>That being said, my active policy is an Expensify US policy, but everytime I go to create an expense while in Spain, it defaults to EUR. 

Are you talking about the currency on the BNP? If so, that's expected as it uses location services / IP address when creating an expense. 

For "P2P distance" here's testing it. 

Personal policy currency is GBP, unit defaults to miles:

<img width="376" alt="Image" src="https://github.com/user-attachments/assets/6403eb00-7aac-4683-ac21-6b5d555ebad6" />

Personal policy currency is EUR, unit defaults to km: 

<img width="374" alt="Image" src="https://github.com/user-attachments/assets/c8a3bdbf-b507-4244-89f5-930629388e4c" />

So I think that's working?

**Comment by shawnborton** [comment]
> Are you talking about the currency on the BNP? If so, that's expected as it uses location services / IP address when creating an expense.

Yeah, that's what I meant - good to hear it's working as intended. 

But yeah, what you are referencing looks like it's working so let's go with that!

**Comment by trjExpensify** [comment]
>But yeah, what you are referencing looks like it's working so let's go with that!

Cool yeah, so the improvement for `Create expense` flow will be to use the unit of your group workspace if that's your activePolicyID. If your activePolicyID is the personal workspace (most likely case is because you haven't created a group workspace yet), then we display the unit being extrapolated from the personal policy output currency as shown above.

With regards to the confirmation screen, what did you think of this? 

>Extra: If you then select a workspace that's different to your default workspace, we can change the unit we display on the confirmation screen preview to match that workspace's setting perhaps?

I think we should do that to ensure that the unit in the rate details in the distance expense matches the unit being displayed in the preview.

**Comment by shawnborton** [comment]
That sounds good to me!

**Comment by dubielzyk-expensify** [comment]
Great suggestions and improvement. 

I actually quite like the ability to tap the marker to change from miles to km. It's one of those things that'd be hidden but a nice little feature for those who find it. Not opposed to leaving it out though. Also it can't mess with the other map gestures that exists today. 

> I think we should do that to ensure that the unit in the rate details in the distance expense matches the unit being displayed in the preview.

Agree with this.

**Comment by shawnborton** [comment]
@truph01 does @trjExpensify's post [here](https://github.com/Expensify/App/issues/53421#issuecomment-2595461198) make sense to you, and is that something you can implement?

**Comment by shawnborton** [comment]
cc @dubielzyk-expensify since you worked on mocks for this - I think the idea is that the tooltip literally just changes from "XX mi" to "XX km" when you tap on it

**Comment by dubielzyk-expensify** [comment]
> cc @dubielzyk-expensify since you worked on mocks for this - I think the idea is that the tooltip literally just changes from "XX mi" to "XX km" when you tap on it

Yep that's correct!

**Comment by dubielzyk-expensify** [comment]
> Do we have any indicator to inform users that they can tap on the label to change the distance unit?

Nope. This is mostly a shortcut for those who know or discover. 

> Additionally, I noticed that in the design image in the OP, the tooltip still displays both distance units. I think we should update the OP to align with the latest design.

I'll update!

**Comment by dubielzyk-expensify** [comment]
Ahh, yes, I'll update the OP but we should follow the latest comment and use the workspace to determine it 👍

**Comment by dubielzyk-expensify** [comment]
Sorry for the confusion 🙏

**Comment by shawnborton** [comment]
New screenshots look good to me. I think we should also add a note to the OP explaining the tap behavior of the distance label as well.

**Comment by dubielzyk-expensify** [comment]
Great shout. I'll update 👍

**Comment by bfitzexpensify** [comment]
@shubham1206agra please complete the BZ checklist

{ayment complete to @truph01

**Comment by bfitzexpensify** [comment]
Bump @shubham1206agra

**Comment by bfitzexpensify** [comment]
Thanks @parasharrajat - looks like both issues tagged the offending PR but not this issue, which is why I didn't see them linked. 

OK, I've adjusted the payment down to $62.50. Offer resent @shubham1206agra

**Comment by bfitzexpensify** [comment]
Not a blocker, but still a regression that had to be addressed.

**Comment by bfitzexpensify** [comment]
Thanks for the context @truph01, I agree given that that we can ignore https://github.com/Expensify/App/issues/56531 as a regression.

**Comment by bfitzexpensify** [comment]
I updated the offer @shubham1206agra - offer should be in your Upwork account

**Comment by bfitzexpensify** [comment]
https://expensify.slack.com/archives/C01SKUP7QR0/p1741026052684239
